### PR TITLE
Add option to move event persistence off master

### PR DESCRIPTION
**Pull request by erikjohnston**
This adds config of the form:

```yaml
writers:
    events: some_worker_name
```

(Better names for config option would be appreciated).

Requires #7515 ~~and #7516~~

### CODE REVIEW

File: synapse/config/workers.py [hunk 1]
```diff
@@ -88,6 +99,21 @@ def read_config(self, config, **kwargs):
             name: InstanceLocationConfig(**c) for name, c in instance_map.items()
         }
 
+        # Map from type of streams to source, c.f. WriterLocations.
+        writers = config.get("writers", {}) or {}
```
Comment (anoadragon453) [inline synapse/config/workers.py:103]:
```suggestion
        writers = config.get("writers") or {}
```

File: synapse/replication/tcp/handler.py [hunk 2]
```diff
@@ -87,6 +89,12 @@ def __init__(self, hs):
                 self._streams_to_replicate.append(stream)
                 continue
 
+            if (
+                isinstance(stream, (EventsStream, BackfillStream))
+                and hs.config.worker.writers.events == hs.get_instance_name()
+            ):
+                self._streams_to_replicate.append(stream)
```
Comment (anoadragon453) [inline synapse/replication/tcp/handler.py:96]: A comment for why we do this would be helpful.

File: synapse/handlers/message.py [hunk 3]
```diff
@@ -888,7 +890,7 @@ async def persist_and_notify_client_event(
 
         This should only be run on master.
```
Comment (anoadragon453) [inline synapse/handlers/message.py:891]: Not sure how our terminology is changing, but I think this docstring should be updated?

File: synapse/storage/data_stores/__init__.py [hunk 4]
```diff
@@ -66,9 +66,9 @@ def __init__(self, main_store_class, hs):
 
                     self.main = main_store_class(database, db_conn, hs)
 
-                    # If we're on a process that can persist events (currently
-                    # master), also instantiate a `PersistEventsStore`
-                    if hs.config.worker.worker_app is None:
+                    # If we're on a process that can persist events also
+                    # instansiate a `PersistEventsStore`
```
Comment (anoadragon453) [inline synapse/storage/data_stores/__init__.py:70]:
```suggestion
                    # instantiate a `PersistEventsStore`
```

File: synapse/storage/data_stores/__init__.py [hunk 5]
```diff
@@ -66,9 +66,9 @@ def __init__(self, main_store_class, hs):
...
+                    if hs.config.worker.writers.events == hs.get_instance_name():
```
Comment (anoadragon453) [inline synapse/storage/data_stores/__init__.py:71]: Ok, so `hs.get_instance_name()` is essentially how we check if we're on a "master" process, correct?

Comment (erikjohnston) [reply synapse/storage/data_stores/__init__.py:71]: Yeah, so each instance has a unique name and we use that to determine which process is in charge of event persistence (by default that is `"master"` but can be configured to be another instance via `writers` config)

File: synapse/replication/tcp/handler.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/replication/tcp/handler.py:98]: Done, let me know if it makes sense

File: synapse/storage/data_stores/__init__.py (hunk 5 above)
Comment (anoadragon453) [reply synapse/storage/data_stores/__init__.py:71]: Makes sense, thanks!

File: synapse/handlers/message.py [hunk 6]
```diff
@@ -886,9 +888,9 @@ async def persist_and_notify_client_event(
         """Called when we have fully built the event, have already
         calculated the push actions for the event, and checked auth.
 
-        This should only be run on master.
+        This should only be run on instance in charge of persisting events.
```
Comment (anoadragon453) [inline synapse/handlers/message.py:891]:
```suggestion
        This should only be run on instances in charge of persisting events.
```

:P

Comment (erikjohnston) [reply synapse/handlers/message.py:891]: nyargh, thanks

**Comment by erikjohnston** [comment]
The commits are mostly me sorting out a bad merge. The test failure is a flakey sytest that I'm dealing with separately.

**Pull request merged**